Framework/Geometry/src/Instrument/InstrumentDefinitionParser.cpp

```diff
@@ -1319,6 +1320,148 @@ void InstrumentDefinitionParser::appendLeaf(Geometry::ICompAssembly *parent,
+    //-------------- Create a StructuredDetector
+    //------------------------------------------------
+    std::string name = InstrumentDefinitionParser::getNameOfLocationElement(
+        pLocElem, pCompElem);
+
+    // Create the bank with the given parent.
+    auto bank = new Geometry::StructuredDetector(name, parent);
+
+    // set location for this newly added comp and set facing if specified in
+    // instrument def. file. Also
+    // check if any logfiles are referred to through the <parameter> element.
+    setLocation(bank, pLocElem, m_angleConvertConst, m_deltaOffsets);
+    setLogfile(
+        bank, pCompElem,
+        m_instrument->getLogfileCache()); // params specified within <component>
+    setLogfile(
+        bank, pLocElem,
+        m_instrument
+            ->getLogfileCache()); // params specified within specific <location>
+
+    // Extract all the parameters from the XML attributes
+    int xpixels = 0;
+    int ypixels = 0;
+    int idstart = 0;
+    bool idfillbyfirst_y = true;
+    int idstepbyrow = 0;
+    int idstep = 1;
+    std::vector<double> xValues;
+    std::vector<double> yValues;
+
+    // The shape!
+    // Given that this leaf component is actually an assembly, its constituent
+    // component detector shapes comes from its type attribute.
+    const std::string shapeType = pType->getAttribute("type");
+    boost::shared_ptr<Geometry::Object> shape = mapTypeNameToShape[shapeType];
+
+    // These parameters are in the TYPE defining StructuredDetector
+    if (pType->hasAttribute("xpixels"))
+      xpixels = atoi((pType->getAttribute("xpixels")).c_str());
+    if (pType->hasAttribute("ypixels"))
+      ypixels = atoi((pType->getAttribute("ypixels")).c_str());
+
+    // THESE parameters are in the INSTANCE of this type - since they will
+    // change.
+    if (pCompElem->hasAttribute("idstart"))
+      idstart = atoi((pCompElem->getAttribute("idstart")).c_str());
+    if (pCompElem->hasAttribute("idfillbyfirst"))
+      idfillbyfirst_y = (pCompElem->getAttribute("idfillbyfirst") == "y");
+    // Default ID row step size
+    if (idfillbyfirst_y)
+      idstepbyrow = ypixels;
+    else
+      idstepbyrow = xpixels;
+    if (pCompElem->hasAttribute("idstepbyrow")) {
+      idstepbyrow = atoi((pCompElem->getAttribute("idstepbyrow")).c_str());
+    }
+    // Default ID row step size
+    if (pCompElem->hasAttribute("idstep"))
+      idstep = atoi((pCompElem->getAttribute("idstep")).c_str());
+
+    // Access type element which defines structured detecor vertices
+    Element *pElem;
+    NodeIterator tags(pCompElem->ownerDocument(), NodeFilter::SHOW_ELEMENT);
+    Node *pNode = tags.nextNode();
+
+    while (pNode) {
+      Element *check = static_cast<Element *>(pNode);
+      if (pNode->nodeName().compare("type") == 0 && check->hasAttribute("is")) {
+        std::string is = check->getAttribute("is").c_str();
+        if (is.compare("StructuredDetector") == 0 ||
```

Comment:
Same comparison issue as above. Push the match responsibility into the `StructuredDetector` maybe something like `static bool compareName(const std::string& proposedMatch);`